checkpoint/recovery_test.go

```diff
@@ -119,33 +120,57 @@ func verifyDbContent(tb testing.TB, db *sql.Database) {
 	require.Empty(tb, extra)
 }
 
-func checkpointServer(t testing.TB) string {
+func checkpointServerWithFaultInjection(t testing.TB, succeed func() bool) string {
 	mux := http.NewServeMux()
 	mux.HandleFunc("GET /snapshot-15", func(w http.ResponseWriter, r *http.Request) {
-		w.Write([]byte(checkpointData))
+		if succeed == nil || succeed() {
+			w.Write([]byte(checkpointData))
+		} else {
+			w.WriteHeader(http.StatusServiceUnavailable)
+			w.Write([]byte("service unavailable"))
+		}
 	})
 	ts := httptest.NewServer(mux)
 	t.Cleanup(ts.Close)
 	return ts.URL
 }
 
+func checkpointServer(t testing.TB) string {
+	return checkpointServerWithFaultInjection(t, nil)
+}
+
 func TestRecover(t *testing.T) {
```

Comment:
In my opinion we should keep tests much simpler than this and avoid sharing setup code with other tests that complicate them too much - generally as soon as an `if` pops up somewhere in a test we should consider writing it a new with less shared logic.

Compare the current implementation of `TestRecover` with this:

```go
func TestRecover(t *testing.T) {
	t.Parallel()

	setup := func(t *testing.T, handler http.HandlerFunc) (afero.Fs, *checkpoint.RecoverConfig) {
		t.Helper()
		fs := afero.NewMemMapFs()

		mux := http.NewServeMux()
		mux.HandleFunc("GET /snapshot-15", handler)
		ts := httptest.NewServer(mux)
		t.Cleanup(ts.Close)
		cfg := &checkpoint.RecoverConfig{
			GoldenAtx:   goldenAtx,
			DataDir:     t.TempDir(),
			DbFile:      "test.sql",
			LocalDbFile: "local.sql",
			NodeIDs:     []types.NodeID{types.RandomNodeID()},
			Uri:         fmt.Sprintf("%s/snapshot-15", ts.URL),
			Restore:     types.LayerID(recoverLayer),
			RetryMax:    5,
			RetryDelay:  100 * time.Millisecond,
		}
		require.NoError(t, fs.MkdirAll(filepath.Join(cfg.DataDir, bootstrap.DirName), 0o700))
		return fs, cfg
	}

	t.Run("http", func(t *testing.T) {
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {
			w.Write([]byte(checkpointData))
		})

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.NoError(t, err)
		require.Nil(t, data)

		newDB, err := sql.Open("file:" + filepath.Join(cfg.DataDir, cfg.DbFile))
		require.NoError(t, err)
		require.NotNil(t, newDB)
		defer newDB.Close()
		verifyDbContent(t, newDB)

		restore, err := recovery.CheckpointInfo(newDB)
		require.NoError(t, err)
		require.EqualValues(t, recoverLayer, restore)
	})

	t.Run("http+retry", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		var fail atomic.Bool
		fail.Store(true)
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {
			if fail.CompareAndSwap(true, false) { // fail on first request
				w.WriteHeader(http.StatusServiceUnavailable)
				w.Write([]byte("service unavailable"))
				return
			}
			w.Write([]byte(checkpointData))
		})

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.NoError(t, err)
		require.Nil(t, data)

		newDB, err := sql.Open("file:" + filepath.Join(cfg.DataDir, cfg.DbFile))
		require.NoError(t, err)
		require.NotNil(t, newDB)
		defer newDB.Close()
		verifyDbContent(t, newDB)

		restore, err := recovery.CheckpointInfo(newDB)
		require.NoError(t, err)
		require.EqualValues(t, recoverLayer, restore)
	})

	t.Run("not found", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {})
		cfg.Uri = strings.Replace(cfg.Uri, "/snapshot-15", "/snapshot-42", -1) // unavailable snapshot

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.ErrorIs(t, err, checkpoint.ErrCheckpointNotFound)
		require.Nil(t, data)
	})

	t.Run("url unreachable", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {})
		cfg.Uri = "http://nowhere/snapshot-15" // unreachable url

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.ErrorIs(t, err, checkpoint.ErrCheckpointRequestFailed)
		require.Nil(t, data)
	})

	t.Run("unsupported scheme", func(t *testing.T) {
		t.Parallel()
		db := sql.InMemory()
		localDB := localsql.InMemory()
		fs, cfg := setup(t, func(w http.ResponseWriter, r *http.Request) {})
		cfg.Uri = "ftp://snapshot-15" // unsupported scheme

		data, err := checkpoint.RecoverWithDb(context.Background(), zaptest.NewLogger(t), db, localDB, fs, cfg)
		require.ErrorIs(t, err, checkpoint.ErrUrlSchemeNotSupported)
		require.Nil(t, data)
	})
}
```

It's in my opinion much clearer what is going on: each test is short, doesn't have any branches and doesn't share any code with tests outside of this specific group. This makes it easier to debug them and when they fail its easier to figure out why they failed instead of going through multiple indirections.

Wdyt?

Reply:
I tend to agree with @fasmat. Ideally if the tests that check for failures start sharing the setup with the tests that test for happy path we end up with something which is less readable and more difficult to maintain. I would suggest to just separate the tests that depend on a mock returning an error to have a separate setup. IMO it is fine to duplicate simple code to get better readability.

Reply:
In general I think there needs some balance, as another thing to consider is that repeating code chunks also obscure the primary purpose of the test, which is in this case is "try to get the URL with or without problem => specific simple outcome"; this purpose being specified concisely in a table basically makes the purpose of the whole test more self-explanatory.
But in this particular case the tests do indeed differ enough from each other, so going with @fasmat's approach is reasonable enough, which I did.